### ISSUE TITLE
Presence of format strings defeats Warning 20 (ignored-extra-argument)

### ISSUE THREAD

**Issue by msprotz** [ocaml/ocaml#12262]
Warning 20 warns when an `'a`-returning function will ignore its argument.

```
# if false then failwith "foo" "bar";;
Warning 20 [ignored-extra-argument]: this argument will not be used by the function.
- : unit = ()
```

This is also works with `let rec ff x = ff x`, `let bye x = print_endline x; exit 0`, i.e. any function that returns an unconstrained type variable.

However, this warning seems to be defeated when using a custom helper that takes a format string *then* fails.

```
# let fatal_error fmt =
    Printf.kbprintf (fun buf ->
      Buffer.add_string buf "\n";
      Buffer.output_buffer stderr buf;
      failwith "Unrecoverable error"
    ) (Buffer.create 16) fmt
  ;;
val fatal_error : ('a, Buffer.t, unit, 'b) format4 -> 'a = <fun>
# if false then fatal_error "%s" "foo" "bar";;
- : unit = ()
```

My expectation is that warning 20 would trigger here, too, since the format4 type indicates that the return type is unconstrained here, too.

**Comment by gasche** [comment]
Fun fact, two ways to add parentheses around some arguments fix the issue:

```ocaml
# if false then (fatal_error "%s" "foo" "bar");; (* bug *)
- : unit = ()
# if false then (fatal_error "%s" "foo") "bar";; (* no bug *)
Warning 20 [ignored-extra-argument]: this argument will not be used by the function.
- : unit = ()
# if false then (fatal_error "%s") "foo" "bar";; (* no bug *)
Warning 20 [ignored-extra-argument]: this argument will not be used by the function.
- : unit = ()
```

This is also not related to format strings directly or their use of GADTs, there are other ways to observe this limitation:

```ocaml
# let apply f x = f x;;
val apply : ('a -> 'b) -> 'a -> 'b = <fun>
# if false then apply failwith "foo" "bar";;
- : unit = ()
# if false then (apply failwith) "foo" "bar";;
Warning 20 [ignored-extra-argument]: this argument will not be used by the function.
- : unit = ()
# if false then (apply failwith "foo") "bar";;
Warning 20 [ignored-extra-argument]: this argument will not be used by the function.
- : unit = ()
```

This seems to be an interaction between the implementation of this warning and the type-checking rule for n-ary applications.

**Comment by gasche** [comment]
So here is the explanation for the issue.

A wrong implementation of the check would be that when we see an application `t u` (where `t` itself may be an application), type-check `t` and, if it is a type variable, emit this warning. This is wrong because it would also fire in the case of unification variables that will later be found to be function types, consider for example just `fun f x -> f x`: at the point where the application `f x` is checked, the type of `f` is an unknown unification variable.

The actual implementation of this check does the following: when checking an application `t u`, first type-check `t` and `x`, but then raise the current inference level from N to N+1 and instantiate the type of `t` (which may be a polymorphic scheme) at this level. At this point we know that "just yet-unknown unification variables" will have level N or less (in our example `f x`, `f` is typed by a variable of level N), while the "generic" parts of the type of `t` will have level N+1. Warn if the type of `t` is a variable at level N+1, but not if it as a lower/older level.
For n-ary applications `t u1 u2 ... un`, instead of doing this recursively, we take an instance `t` at a higher level, and then check all partial applications with this criterion.

As a result, `failwith "foo" "bar"` warns because the instance of `failwith` at level N+1 has type `string -> 'a`, with `'a` at level N+1. On the other hand, `apply failwith "foo" "bar"` first type-checks `failwith` at level N, so we get a first argument of `apply` at type `string -> 'b` with `'b` at level N. This later gets unified with the type of the partial application `apply failwith "foo"`, but the warning does not fire at this point because the variable has level N, not N+1.

In other words: if we wanted this warning to work well for application of higher-order functions, I suppose that we could try to instantiate not just the type of the function at level N+1, but also the type of the arguments. Or maybe we could "just" recursively raise the level and instantiate for each partial application -- but this could introduce quadratic behaviors and may have a negative impact on the type-checking time, as I suppose that this logic is fairly hot.

**Comment by jberdine** [comment]
While not a solution to the core issue, a practical work-around that can be done in the language now is to change functions like `fatal_error` to have types such as `val fatal_error : ('a, Buffer.t, unit, unit -> 'b) format4 -> 'a` by changing `Printf.kbprintf (fun buf ->` to `Printf.kbprintf (fun buf () ->`. Changing the return type from `'b` to `unit -> 'b` requires calls to `fatal_error` to pass an extra `()` argument just after the arguments consumed by the format string. This makes it clear at call sites which arguments are consumed by the format string, and leads to passing too many or few into type errors. In my experience this approach is robust and not too heavy.

**Comment by Octachron** [comment]
My favorite side-solution is to have a version of `Format` interpreters that takes a heterogeneous list as an argument. This avoids both the need for the `k` continuation variants and the complex error messages when the number of arguments doesn't match the format string.

**Comment by jberdine** [comment]
> My favorite side-solution is to have a version of `Format` interpreters that takes a heterogeneous list as an argument.

I'd be interested to see an example of this approach if you have a pointer handy.

**Comment by Octachron** [comment]
With a [hlist](https://gist.github.com/Octachron/6d503f5edaad633c4736c222d25e4a61) as argument, we can write the `fatal_error` function as:
```ocaml
let fatal_error fmt args =
  let ppf = Format.err_formatter in
  fprintf ppf fmt args;
  fprintf ppf "@." [];
  failwith "Unrecoverable error"
```
Then, the example in this PR can be written as
```ocaml
let test = if false then fatal_error "%s" ["foo";"bar"]
```
and fails with a quite sibylline error
```ocaml
22 | let test = if false then fatal_error "%s" ["foo";"bar"]
                                                      ^^^^^^
Error: This expression has type ('a -> 'b) L.t
       but an expression was expected of type unit L.t
       Type 'a -> 'b is not compatible with type unit 
```
(which is in fact stating that we were expecting a list of size one rather than two.)
Nevertheless, the real benefit here is that the error message is correctly pointing at the supernumerary list element `"bar"`.

**Comment by gasche** [comment]
This is a nice trick, should we consider including it in the standard library?

(This could be exposed as a wrapper after the fact, or hidden as new variants of the `printf` function, maybe `l{,s,k}printf` or something. Having new variants would let migrate to a variant of `CamlinternalFormat.make_printf` that takes the heterogeneous list directly, which would be slightly more efficient than what we currently have.)

**Comment by Octachron** [comment]
I fear that the issue would be more on introducing one definition of heterogeneous list amongst of the many possible versions of heterogeneous list in the standard library which would open numerous venues for bikeshedding.

**Comment by gasche** [comment]
Introducing it in CamlinternalFormat and relying on type-directed lookup could be a way forward, by very intentionally not trying to provide a one-size-fits-all solution.

**Comment by Octachron** [comment]
I am more and more convinced that having a function in `Format` which takes a heterogeneous list as arguments would solve many of those usability issues.

**Comment by gasche** [comment]
This sounds like a good L3/M1-level internship topic to me, with a useful goal and a large amount of GADT programming in perspective.

**Comment by zazedd** [comment]
I decided to work on this!
I currently group all the new functions inside a module, but we may prefer the `l{,s,k}printf` naming convention mentioned by @gasche.
It is still incomplete as I have not tested everything thoroughly (that's why `kdprintf`is commented out), but I wanted to get an opinion from all of you guys to make sure that the way I am currently implementing this is okay.

**Comment by zazedd** [comment]
Also, my solution has the bonus of fixing this issue:

```ocaml
# Format.ksprintf failwith "error: %i" 42 "unused";; (* bug *)
Exception: Failure "error: 42".

# Format.HetList.ksprintf failwith "error: %i" [ 42; "unused" ];;
Error: The constructor [] has type ('a, 'a) Format.HetList.t
       but an expression was expected of type
         ('a, string -> 'a) Format.HetList.t
       The type variable 'a occurs inside string -> 'a
```
The error in fact states that there is an extra `string` argument there.

**Comment by Octachron** [comment]
One of the advantage of the heterogeneous version is that less variants are needed. A version of `printf`, `eprintf` ,`fprintf` and `asprintf` should be enough.

The continuation versions are useful for the "variadic" interpreters because there is no syntactic notion on when the interpretation of the format string is finished, this no longer the case when using a delimited list of arguments. Similarly, the `dprintf` cannot be easily derived from the `fprintf` version in the continuation form, but that is not the case in the list form.

**Comment by zazedd** [comment]
That is true!
I removed the continuation variants and the `dprintf` function. I also moved the functions to the outside of the new `HetList` module because they are now few enough that the naming scheme `l{,s,k}printf` works nicely.
I also added some documentation to the new functions and module. I would love some feedback on the latest changes.

I'm unsure if it is okay that `HetList.apply` is exposed. We could move it to `CamlinternalFormat` instead.
I'm also not sure if it would be worth having this `HetList` module that only contains the heterogeneous list type definition. To be clear, I am currently doing it this way because there are tons of `open Format` statements out there, and I do not want to shadow their default list constructors with these. It is also nice to write the heterogeneous lists as normal ones, and this way, we have the best of both worlds.

I would also like to know if `@since 5.2.1` is correct :)

**Comment by Octachron** [comment]
The probable version would probably be more `5.4` since we are getting close to the feature freeze for `5.3`  (see [introduction.md](../blob/trunk/release-info/introduction.md) for the meaning of version number).

The name `HetList` is probably too generic, I would rather propose `Arg_list`.

In my opinion,  the `apply` function is an implementation detail that doesn't need to be exposed in the interface.

**Comment by zazedd** [comment]
Sure thing!

The `apply` function will have to move with the new `Arg_list` module to not cause cyclic dependencies. Looking at the code in `CamlinternalFormat` and `CamlinternalFormatBasics`, I think the new module should go into `CamlinternalFormat`.
 However, this module is not exposed to `Format`'s `mli` file and so I can't properly type the new functions there. `CamlinternalFormatBasics` is exposed, and everything works when the module is inside of it. What do you reckon I should do in this case?

**Comment by Octachron** [comment]
I am not sure which issue you see with defining the `Arg_list` module inside `Format`. At most, it would shadow other modules names `Arg_list` after opening the `Format` module.

**Comment by gasche** [comment]
I would expect `Printf` to also get such an interface eventually. I think that it is reasonable to expect the same argument-list definition to work for both modules. This would suggest going to CamlinternalFormat, the shared parent, instead of Format.

**Comment by zazedd** [comment]
[Previously](https://github.com/zazedd/ocaml/commit/ed0579c3f7f5cc05501bec2201816091a47e001e), the `HetList` module was inside the `Format` module, and the `apply` function was accessible in `Format.HetList.apply`, so it was technically exposed.
Because the `apply` function depends on the heterogeneous list definition, I thought to move the whole module to another place, in this case `CamlinternalFormat` or `CamlinternalFormatBasics`. Probably no one other than the Stdlib will actually open any of these, so there wouldn't be any shadowing :)

**Comment by zazedd** [comment]
> I would expect Printf to also get such an interface eventually. I think that it is reasonable to expect the same argument-list definition to work for both modules. This would suggest going to CamlinternalFormat, the shared parent, instead of Format.

That makes total sense, I will look into it!

**Comment by Octachron** [comment]
Scanf is another potential secondary user but with less clear benefits since it would produce a heterogeneous list rather than consume a short-lived literal.

**Comment by gasche** [comment]
I still see benefits in writing something like `scanf ibuf (fun [foo; bar; baz] -> ...)`, but I don't know if people actually use Scanf nowadays except for throwaway code, so I'm not sure it should be on the priority list at all.

**Comment by Octachron** [comment]
I was more thinking of writing `let [foo;bar;baz] = Scanf.scanf "%d.%d.%d" ` , but yes this is at most an afterthought let for later.

**Comment by zazedd** [comment]
In preparation for new heterogeneous list variants in `Printf`, I created a new function called `make_lprintf` inside `CamlinternalFormat`. It does the work of `make_printf` but applies the heterogeneous list arguments right after.

The `Arg_list` module, with the type definition and `apply` function of the heterogeneous list is now in `CamlinternalFormatBasics`.

> I was more thinking of writing let [foo;bar;baz] = Scanf.scanf "%d.%d.%d"

This is pretty cool and, in my opinion, more user-friendly than what `Scanf.scanf` does right now.
I will look into it as well :).

**Comment by zazedd** [comment]
I changed the location of `Arg_list` from `CamlinternalFormatBasics` to `CamlinternalFormat`, but I had to change `format.mli`'s `.depend` part to have `camlinternalFormat.cmi`, I don't know if that's okay :)

---

As for the `Printf` variations, `make_lprintf` is one way to do it. `Printf` functions would look like this (similar to the current new `Format` functions):

```ocaml
let lfprintf : type a .
  out_channel ->
  (a, out_channel, unit) format ->
  (a, unit) Arg_list.t ->
  unit =
  fun oc (Format (fmt, _)) args ->
  make_lprintf (fun acc -> output_acc oc acc)
    End_of_acc fmt args

let lprintf : type a .
  (a, out_channel, unit) format ->
  (a, unit) Arg_list.t ->
  unit =
  fun fmt args ->
  lfprintf stdout fmt args
  
...
```

Which introduces _some_ repetition.

Another option is to do what I was previously doing for `Format`. Because now we have `Arg_list` and the `apply` function available inside `CamlinsideFormat`, we can just do this:

```ocaml
let lfprintf : type a .
  out_channel ->
  (a, out_channel, unit) format ->
  (a, unit) Arg_list.t ->
  unit =
  fun oc fmt args ->
  let f = fprintf oc fmt in
  Arg_list.apply f args

let lprintf : type a .
  (a, out_channel, unit) format ->
  (a, unit) Arg_list.t ->
  unit =
  fun fmt args ->
  lfprintf stdout fmt args
  
...
```

Which is clean! But to be fair the other way is also pretty fine, and we are not relying on `kfprintf` (as ```let fprintf oc fmt = kfprintf ignore oc fmt```). 

Tell me which way you prefer.

**Comment by zazedd** [comment]
Just added `Scanf.lscanf`, which can be used as such:

```ocaml
let Scanf.Arg_list.[ x1; x2 ] = Scanf.lscanf "%s %d";;
string 123
val x1 : string = "string"
val x2 : int = 123 

```

Or something like:
```ocaml
match Scanf.lscanf "%s %d" with 
  | [ x1; x2 ] -> ... 
  | _ -> .
```

I don't know if the import for `Scanf.Arg_list` can be removed (I haven't found a way to do so yet) without opening it first and overloading the default list constructors, but it would be great.

I currently expose `CamlinternalFormat.Arg_list` through `Scanf`, so users don't have to write `let CamlinternalFormat.Arg_list.[ x1; x2 ]`, as they aren't supposed to interact with that module directly.

---

Regarding my previous comment, I decided to do it like the first code block. With this commit, I added new `Printf` variants.
Whenever the presented code is ready, I can split it into multiple PRs so that the new features for each new module are separated.

**Comment by nojb** [comment]
> Just added `Scanf.lscanf`, which can be used as such:

Not been following the discussion in detail, but my feeling is that it may be better to only expose an option-returning variant of the new function. Which also has the consequence of forcing the use of `match` on the result, which enables eliding the local open via type-based disambiguation of constructors. (But maybe I am missing some GADT subtleties.)

**Comment by zazedd** [comment]
> it may be better to only expose an option-returning variant of the new function

Sounds like a good idea!

I am going to start the pull requests, the first one will introduce the new `Format` functions and the list definition.

**Comment by jberdine** [comment]
Feel free to ignore this, but let me mention something that I think would be an improvement while this is being thought about. I don't know if it is possible, but I think it would help readability and maybe errors if the heterogeneous list could be tweaked so that formatting functions could be paired with their arguments. For example, being able to write `fatal_error "%s%a" ["foo"; (pp_print_string, "bar")]` instead of `fatal_error "%s%a" ["foo"; pp_print_string; "bar"]`. I haven't been able to think of a typing technique that would work for this, but maybe others know how?

**Comment by zazedd** [comment]
> Feel free to ignore this, but let me mention something that I think would be an improvement while this is being thought about. I don't know if it is possible, but I think it would help readability and maybe errors if the heterogeneous list could be tweaked so that formatting functions could be paired with their arguments. For example, being able to write `fatal_error "%s%a" ["foo"; (pp_print_string, "bar")]` instead of `fatal_error "%s%a" ["foo"; pp_print_string; "bar"]`. I haven't been able to think of a typing technique that would work for this, but maybe others know how?

Sorry, I did not see this reply before creating the PR :(
I will look into that, although I don't see a way off the top of my head either.

**Comment by nojb** [comment]
Closing this as the heterogeneous list proposal was merged in #13372.
